cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,951 @@
+
+template <typename vertex_t, typename edge_t, typename EdgeBuffer>
+struct unroll_edge {
+  edge_t num_valid_edges{};
+  raft::device_span<edge_t> num_triangles{};
+  EdgeBuffer edge_to_unroll_first{};
+  EdgeBuffer transposed_valid_edge_first{};
+  EdgeBuffer transposed_valid_edge_last{};
+  EdgeBuffer transposed_invalid_edge_last{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    // edges are sorted with destination as key so reverse the edge when looking it
+    auto pair = thrust::make_tuple(thrust::get<1>(*(edge_to_unroll_first + i)),
+                                   thrust::get<0>(*(edge_to_unroll_first + i)));
+    // Find its position in either partition of the transposed edgelist
+    // An edge can be in found in either of the two partitions (valid or invalid)
+
+    auto itr = thrust::lower_bound(
+      thrust::seq, transposed_valid_edge_last, transposed_invalid_edge_last, pair);
+    auto idx = thrust::distance(transposed_valid_edge_last, itr) + num_valid_edges;
+
+    if (*itr != pair) {
+      // The edge must be in the first boundary
+      itr = thrust::lower_bound(
+        thrust::seq, transposed_valid_edge_first, transposed_valid_edge_last, pair);
+      idx = thrust::distance(transposed_valid_edge_first, itr);
+    }
+
+    assert(*itr == pair);
+    cuda::atomic_ref<edge_t, cuda::thread_scope_device> atomic_counter(num_triangles[idx]);
+    auto r = atomic_counter.fetch_sub(edge_t{1}, cuda::std::memory_order_relaxed);
+  }
+};
+
+template <typename vertex_t>
+rmm::device_uvector<vertex_t> prefix_sum_valid_and_invalid_edges(
+  // The edgelist is segmented into 2 partitions (valid and invalid edges)
+  // and edges to be unrolled can be either in the valid or the invalid edge
+  // partition.
+  raft::handle_t const& handle,
+  raft::device_span<vertex_t const> invalid_dsts,
+  raft::device_span<vertex_t const> edgelist_dsts)
+{
+  rmm::device_uvector<vertex_t> prefix_sum(invalid_dsts.size() + 1, handle.get_stream());
+
+  thrust::tabulate(handle.get_thrust_policy(),
+                   prefix_sum.begin(),
+                   prefix_sum.begin() + invalid_dsts.size(),
+                   [invalid_dsts,
+                    num_edges     = edgelist_dsts.size(),
+                    edgelist_dsts = edgelist_dsts.begin()] __device__(auto idx) {
+                     auto itr_lower_valid = thrust::lower_bound(
+                       thrust::seq, edgelist_dsts, edgelist_dsts + num_edges, invalid_dsts[idx]);
+
+                     auto itr_upper_valid = thrust::upper_bound(
+                       thrust::seq, itr_lower_valid, edgelist_dsts + num_edges, invalid_dsts[idx]);
+
+                     auto dist_valid = thrust::distance(itr_lower_valid, itr_upper_valid);
+
+                     return dist_valid;
+                   });
+  thrust::exclusive_scan(
+    handle.get_thrust_policy(), prefix_sum.begin(), prefix_sum.end(), prefix_sum.begin());
```

Comment:
You can merge this into a single scan call.
```
auto count_first = thrust::make_transform_iterator(thrust::make_counting_iterator(size_t{0}), [edgelist_dsts, invalid_dsts]__device__(size_t i) {
  ...;
});
thrust::exclusive_scan(handle.get_thrust_policy(), count_first, count_first + ..., prefix_sum.begin());
```

Reply:
Somehow I can't make this to work. Perhaps we can sync regarding this.